src/Compilers/CSharp/Portable/Symbols/Source/SourceModuleSymbol.cs

```diff
@@ -322,6 +335,79 @@ private void ValidateLinkedAssemblies(BindingDiagnosticBag diagnostics, Cancella
             }
         }
 
+        internal void DiscoverInterceptorsIfNeeded()
+        {
+            if (_state.NotePartComplete(CompletionPart.StartInterceptorsChecks))
+            {
+                DiscoverInterceptors();
+                _state.NotePartComplete(CompletionPart.FinishInterceptorsChecks);
+            }
+        }
+
+        private void DiscoverInterceptors()
+        {
+            var toVisit = ArrayBuilder<NamespaceOrTypeSymbol>.GetInstance();
+
+            var location = this.GlobalNamespace.GetFirstLocationOrNone();
+            if (!location.IsInSource)
+            {
+                return;
+            }
+
+            // Search the namespaces which were indicated to contain interceptors.
+            var interceptorsNamespaces = ((CSharpParseOptions)location.SourceTree.Options).InterceptorsPreviewNamespaces;
+            foreach (var namespaceParts in interceptorsNamespaces)
+            {
+                if (namespaceParts is ["global"])
+                {
+                    toVisit.Add(GlobalNamespace);
+                    // No point in continuing, we already are going to search the entire module in this case.
+                    break;
+                }
+
+                var cursor = GlobalNamespace;
+                foreach (var namespacePart in namespaceParts)
+                {
+                    cursor = (NamespaceSymbol?)cursor.GetMembers(namespacePart).FirstOrDefault(member => member.Kind == SymbolKind.Namespace);
+                    if (cursor is null)
+                    {
+                        break;
+                    }
+                }
+
+                if (cursor is not null)
+                {
+                    toVisit.Add(cursor);
+                }
+            }
+
+            while (toVisit.Count > 0)
+            {
+                var item = toVisit.Pop();
+                if (item is SourceMemberContainerTypeSymbol type)
+                {
+                    type.DiscoverInterceptors(toVisit);
+                }
+                else if (item is SourceNamespaceSymbol @namespace)
+                {
+                    foreach (var member in @namespace.GetMembers())
+                    {
+                        if (member is not NamespaceOrTypeSymbol namespaceOrType)
+                        {
+                            throw ExceptionUtilities.UnexpectedValue(member);
+                        }
+
+                        toVisit.Add(namespaceOrType);
+                    }
+                }
+                else
+                {
+                    throw ExceptionUtilities.UnexpectedValue(item);
+                }
+            }
+        }
```

Comment:
>} [](http://example.com/codeflow?start=8&length=1)

Once we reach here, it would be good to set some debug flag to true and assert that we are not adding new values to ```_interceptions``` when the flag is set.